### ISSUE TITLE
dhcp4: data race in TestPortableConn and TestLinuxConn

### ISSUE THREAD

**Issue by mikioh** [google/netboot#40]
Just tried dhcp4 with go tip (equivalent to go1.9, sorry for the inconvenience of golang/go#19209) and saw it.
```
=== RUN   TestPortableConn
==================
WARNING: DATA RACE
Write at 0x00c42000c620 by goroutine 6:
  go.universe.tf/netboot/dhcp4.testConn()
      /home/mikioh/NFS/src/go.universe.tf/netboot/dhcp4/conn_test.go:73 +0x750
  go.universe.tf/netboot/dhcp4.TestPortableConn()
      /home/mikioh/NFS/src/go.universe.tf/netboot/dhcp4/conn_test.go:133 +0x1ca
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:610 +0xc9

Previous read at 0x00c42000c620 by goroutine 7:
  go.universe.tf/netboot/dhcp4.testConn.func1()
      /home/mikioh/NFS/src/go.universe.tf/netboot/dhcp4/conn_test.go:57 +0x3b

Goroutine 6 (running) created at:
  testing.(*T).Run()
      /usr/lib/golang/src/testing/testing.go:646 +0x52f
  testing.RunTests.func1()
      /usr/lib/golang/src/testing/testing.go:793 +0xb9
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:610 +0xc9
  testing.RunTests()
      /usr/lib/golang/src/testing/testing.go:799 +0x4ea
  testing.(*M).Run()
      /usr/lib/golang/src/testing/testing.go:743 +0x12f
  main.main()
      go.universe.tf/netboot/dhcp4/_test/_testmain.go:76 +0x1b8

Goroutine 7 (finished) created at:
  go.universe.tf/netboot/dhcp4.testConn()
      /home/mikioh/NFS/src/go.universe.tf/netboot/dhcp4/conn_test.go:58 +0x3c7
  go.universe.tf/netboot/dhcp4.TestPortableConn()
      /home/mikioh/NFS/src/go.universe.tf/netboot/dhcp4/conn_test.go:133 +0x1ca
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:610 +0xc9
==================
```
```
=== RUN   TestLinuxConn
==================
WARNING: DATA RACE
Write at 0x00c42000ca20 by goroutine 10:
  go.universe.tf/netboot/dhcp4.testConn()
      /home/mikioh/NFS/src/go.universe.tf/netboot/dhcp4/conn_test.go:73 +0x750
  go.universe.tf/netboot/dhcp4.TestLinuxConn()
      /home/mikioh/NFS/src/go.universe.tf/netboot/dhcp4/conn_test.go:155 +0x1a7
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:610 +0xc9

Previous read at 0x00c42000ca20 by goroutine 11:
  go.universe.tf/netboot/dhcp4.testConn.func1()
      /home/mikioh/NFS/src/go.universe.tf/netboot/dhcp4/conn_test.go:57 +0x3b

Goroutine 10 (running) created at:
  testing.(*T).Run()
      /usr/lib/golang/src/testing/testing.go:646 +0x52f
  testing.RunTests.func1()
      /usr/lib/golang/src/testing/testing.go:793 +0xb9
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:610 +0xc9
  testing.RunTests()
      /usr/lib/golang/src/testing/testing.go:799 +0x4ea
  testing.(*M).Run()
      /usr/lib/golang/src/testing/testing.go:743 +0x12f
  main.main()
      go.universe.tf/netboot/dhcp4/_test/_testmain.go:76 +0x1b8

Goroutine 11 (finished) created at:
  go.universe.tf/netboot/dhcp4.testConn()
      /home/mikioh/NFS/src/go.universe.tf/netboot/dhcp4/conn_test.go:58 +0x3c7
  go.universe.tf/netboot/dhcp4.TestLinuxConn()
      /home/mikioh/NFS/src/go.universe.tf/netboot/dhcp4/conn_test.go:155 +0x1a7
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:610 +0xc9
==================
```

**Comment by danderson** [comment]
Thanks for the report!

How are you getting dhcp4 to run successfully with tip? I just get the tests failing with:

```
--- FAIL: TestPortableConn (0.00s)
	conn_test.go:130: creating the conn: operation not supported
--- FAIL: TestLinuxConn (0.00s)
	conn_test.go:152: creating the linuxconn: operation not supported
```

Do you have in-flight changes to repair the plumbing of x/net, or is my tip install somehow borked?

**Comment by danderson** [comment]
Ah, never mind, I'm just an idiot and forgot about vendored dependencies. I'm able to reproduce this on Go 1.8.1 as well, looking into it.

**Comment by danderson** [comment]
Closed by e9e2bb6, which I forgot to cross-link to this bug before pushing. Doh.
